### DISCUSSION TITLE
Usage of soundfont drumkits?

### DISCUSSION THREAD

**Discussion by henrikvilhelmberglund** [danigb/smplr#47] [Q&A] [answered]
Hi, loving this library! 

I would like to use soundfont drumkits (aka midi channel 10 banks). It seems like these are missing from the prerendered soundfonts. Is there any other way to add them?

**Comment by danigb** [comment]
As far as I know there's no soundfont drumkits available out there. I'm open to support them if you find it. My idea is to support, at some point, something like https://github.com/sfzinstruments/virtuosity_drums but it will take time until sfz file format is fully supported.

**Comment by henrikvilhelmberglund** [comment]
I think most GM soundfonts have drumkits included. FluidR3_GM does for sure:
```
128:0 = Standard
128:1 = Standard 1
128:2 = Standard 2
128:3 = Standard 3
128:4 = Standard 4
128:5 = Standard 5
128:6 = Standard 6
128:7 = Standard 7
128:8 = Room
128:9 = Room 1
128:10 = Room 2
128:11 = Room 3
128:12 = Room 4
128:13 = Room 5
128:14 = Room 6
128:15 = Room 7
128:16 = Power
128:17 = Power 1
128:18 = Power 2
128:19 = Power 3
128:24 = Electronic
128:25 = TR-808
128:32 = Jazz
128:33 = Jazz 1
128:34 = Jazz 2
128:35 = Jazz 3
128:36 = Jazz 4
128:40 = Brush
128:41 = Brush 1
128:42 = Brush 2
128:48 = Orchestra Kit
```
so I guess the problem is that they are not hosted anywhere. I think I could extract them and put them up somewhere. What should I keep in mind to make using them in smplr easier?

**Reply by danigb** [reply]
Exactly, the problem is hosting. If you extract and put it somewhere, we can use it. Currently, hosted Soundfonts uses MIDI.js audio format (see other hosting repositories). It would be nice to use the same, if possible

**Reply by henrikvilhelmberglund** [reply]
I have started on this, I have a repo with some .mp3 files [here](https://github.com/henrikvilhelmberglund/midi-js-compat-soundfonts/tree/gh-pages/GM-soundfonts/FluidR3_GM/drumkits/Standard/mp3) and will try to create a .js with the base64 strings and the correct format. I don't really know how it will work though since drumkits have a lot of missing notes compared to normal GM instruments.

Also not sure what the .mp3 quality settings should be, I think now it is about 220kbps which may be overkill.

This is my workflow so far:
1. Run a node script using midi-writer-js to get .mid files for each note (channel 10 to make sure we get the drumkit)
```js
import MidiWriter from "midi-writer-js";
import fs from "fs";

const octaves = [1, 2, 3, 4, 5, 6];
const notes = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];

octaves.forEach((octave) => {
	notes.forEach((note) => {
		const track = new MidiWriter.Track();
		track.addEvent(new MidiWriter.NoteEvent({ pitch: [`${note}${octave}`], duration: "8", velocity: 100, channel: 10 }));
		fs.writeFileSync(`src/lib/utilities/sf2tobase64/output/${note}${octave}.mid`, new MidiWriter.Writer(track).buildFile());
	});
});
```
2. Import the .mid files in [Reaper](https://www.reaper.fm/) on a single track.
3. Add [BassMIDI VSTi](https://www.kvraudio.com/product/bassmidi-vsti-by-falcosoft) to the track and drag and drop the original soundfont file onto it.
4. Select all items and `Render` changing source to `Selected media items` with `$item` as the filename. Set a directory, the file type to `.wav` and `Tail` to 5000ms or so.
5. In a new project import the `.wavs` on the same track. (The following steps are to remove the extra silence)
6. Select all of the files and do `Item - Item processing - Dynamic split items.
7. Split points `When gate closes` only. `Min slice length 20ms`, `Min silence length 65ms`, `Threshold -70.0 dB` and make sure `Remove silent areas` is ticked at the bottom.
8. After splitting I deleted any unused splits but I guess this could be done after rendering as well by deleting the files that don't have note names.
9. `Render` changing source to `Selected media items` with `$item` as the filename. Set a directory and the file type to `.mp3`.
TODO: create the .js file with the .mp3s as base64 strings.

**Reply by danigb** [reply]
I never did it before. Can't help here.

**Reply by henrikvilhelmberglund** [reply]
No problem, I think I can figure something out!
The plan was to do it a bit faster but my internet stopped working which makes things a bit difficult... I will reply again when I manage to upload something that looks like it should work in smplr ☺️

**Reply by henrikvilhelmberglund** [reply, marked as answer]
I have a drumkit here now:
https://github.com/henrikvilhelmberglund/midi-js-compat-soundfonts/tree/gh-pages/GM-soundfonts/FluidR3_GM/drumkits
It seems like it works if I pass the link in `instrumentUrl`. It says `MIDI.Soundfont.marimba = {` though, not sure what to do about that since it's not really a GM instrument but maybe it doesn't matter? Anyway, seems like I can add new drumkits and use them as well which is nice.

I have an online demo here as well: https://keyboard.henrikvilhelmberglund.com/soundfont-drums

**Reply by danigb** [reply]
Nice!! Congrats 🎉  How do you think we should integrate that into smplr?

**Reply by henrikvilhelmberglund** [reply]
Thanks! 😊 That's a good question... I noticed some discussion here https://github.com/gleitz/midi-js-soundfonts/issues/32 about drumkits as well but not sure what the best option is.
